File: frontend/src/services/vfs.ts
import { api } from "../api/client";
import type { ResolvedVfsSurface, ResolvedVfsSurfaceSource } from "../types/context";

// ─── Descriptor（能力发现） ─────────────────────────────

export interface SelectorHint {
  trigger: string;
  placeholder: string;
  result_item_type: string;
}

export interface VfsDescriptor {
  id: string;
  label: string;
  kind: string;
  provider: string;
  supports: string[];
  root?: string | null;
  workspace_id?: string | null;
  selector?: SelectorHint | null;
}

export interface ListVfssResponse {
  spaces: VfsDescriptor[];
}

// ─── Entry（条目搜索） ──────────────────────────────────

export interface VfsEntry {
  address: string;
  label: string;
  entry_type: string;
  size?: number | null;
  is_dir?: boolean | null;
}

export interface ListEntriesResponse {
  entries: VfsEntry[];
}

// ─── Surface Mount Entry ───────────────────────────────

export interface SurfaceMountEntry {
  path: string;
  entry_type: string;
  size?: number | null;
  is_dir: boolean;
}

export interface ListSurfaceMountEntriesResponse {
  surface_ref: string;
  mount_id: string;
  entries: SurfaceMountEntry[];
}

export interface ReadSurfaceFileResponse {
  surface_ref: string;
  mount_id: string;
  path: string;
  content: string;
  size: number;
}

export interface WriteSurfaceFileResponse {
  surface_ref: string;
  mount_id: string;
  path: string;
  size: number;
  persisted: boolean;
}

export interface ApplySurfacePatchResponse {
  surface_ref: string;
  mount_id: string;
  added: string[];
  modified: string[];
  deleted: string[];
}

export interface VfsQueryParams {
  workspaceId?: string | null;
}

export interface ListEntriesParams extends VfsQueryParams {
  query?: string;
  path?: string;
  recursive?: boolean;
}

function applyQueryParams(searchParams: URLSearchParams, params?: VfsQueryParams) {
  if (!params) return;
  if (params.workspaceId) searchParams.set("workspace_id", params.workspaceId);
}

function buildQs(searchParams: URLSearchParams): string {
  const qs = searchParams.toString();
  return qs ? `?${qs}` : "";
}

export async function listVfss(
  params?: VfsQueryParams,
): Promise<ListVfssResponse> {
  const sp = new URLSearchParams();
  applyQueryParams(sp, params);
  return api.get<ListVfssResponse>(`/vfs${buildQs(sp)}`);
}

export async function listAddressEntries(
  spaceId: string,
  params?: ListEntriesParams,
): Promise<ListEntriesResponse> {
  const sp = new URLSearchParams();
  applyQueryParams(sp, params);
  if (params?.query) sp.set("query", params.query);
  if (params?.path) sp.set("path", params.path);
  if (params?.recursive !== undefined) sp.set("recursive", String(params.recursive));

  return api.get<ListEntriesResponse>(
    `/vfs/${encodeURIComponent(spaceId)}/entries${buildQs(sp)}`,
  );
}

export async function resolveVfsSurface(
  source: ResolvedVfsSurfaceSource,
): Promise<ResolvedVfsSurface> {
  return api.post<ResolvedVfsSurface>("/vfs-surfaces/resolve", { source });
}

export async function getVfsSurface(surfaceRef: string): Promise<ResolvedVfsSurface> {
  return api.get<ResolvedVfsSurface>(`/vfs-surfaces/${encodeURIComponent(surfaceRef)}`);
}

export async function listSurfaceMountEntries(params: {
  surfaceRef: string;
  mountId: string;
  path?: string;
  pattern?: string;
  recursive?: boolean;
}): Promise<ListSurfaceMountEntriesResponse> {
  const sp = new URLSearchParams();
  if (params.path) sp.set("path", params.path);
  if (params.pattern) sp.set("pattern", params.pattern);
  if (params.recursive !== undefined) sp.set("recursive", String(params.recursive));

  return api.get<ListSurfaceMountEntriesResponse>(
    `/vfs-surfaces/${encodeURIComponent(params.surfaceRef)}/mounts/${encodeURIComponent(params.mountId)}/entries${buildQs(sp)}`,
  );
}

export async function readSurfaceFile(params: {
  surfaceRef: string;
  mountId: string;
  path: string;
}): Promise<ReadSurfaceFileResponse> {
  return api.post<ReadSurfaceFileResponse>("/vfs-surfaces/read-file", {
    surface_ref: params.surfaceRef,
    mount_id: params.mountId,
    path: params.path,
  });
}

export async function writeSurfaceFile(params: {
  surfaceRef: string;
  mountId: string;
  path: string;
  content: string;
}): Promise<WriteSurfaceFileResponse> {
  return api.post<WriteSurfaceFileResponse>("/vfs-surfaces/write-file", {
    surface_ref: params.surfaceRef,
    mount_id: params.mountId,
    path: params.path,
    content: params.content,
  });
}

export async function applySurfacePatch(params: {
  surfaceRef: string;
  mountId: string;
  patch: string;
}): Promise<ApplySurfacePatchResponse> {
  return api.post<ApplySurfacePatchResponse>("/vfs-surfaces/apply-patch", {
    surface_ref: params.surfaceRef,
    mount_id: params.mountId,
    patch: params.patch,
  });
}

File: crates/agentdash-contracts/src/vfs.rs
use serde::{Deserialize, Serialize};
use ts_rs::TS;

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
pub struct SelectorHint {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(optional)]
    pub trigger: Option<String>,
    pub placeholder: String,
    pub result_item_type: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
pub struct VfsDescriptor {
    pub id: String,
    pub label: String,
    pub kind: String,
    pub provider: String,
    pub supports: Vec<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(optional)]
    pub selector: Option<SelectorHint>,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
pub struct ListVfssResponse {
    pub spaces: Vec<VfsDescriptor>,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
pub struct VfsEntry {
    pub address: String,
    pub label: String,
    pub entry_type: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(optional)]
    #[ts(type = "number")]
    pub size: Option<u64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(optional)]
    pub is_dir: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
pub struct ListEntriesResponse {
    pub entries: Vec<VfsEntry>,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
pub struct ConfigurableProviderInfo {
    pub service_id: String,
    pub display_name: String,
    pub root_ref_hint: String,
    pub supported_capabilities: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS, PartialEq, Eq)]
#[serde(tag = "source_type", rename_all = "snake_case")]
pub enum ResolvedVfsSurfaceSource {
    ProjectPreview {
        project_id: String,
    },
    StoryPreview {
        project_id: String,
        story_id: String,
    },
    TaskPreview {
        project_id: String,
        task_id: String,
    },
    SessionRuntime {
        session_id: String,
    },
    ProjectSkillAssets {
        project_id: String,
    },
    ProjectVfsMount {
        project_id: String,
        mount_id: String,
    },
    ProjectAgentKnowledge {
        project_id: String,
        project_agent_id: String,
    },
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, TS, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum ResolvedMountPurpose {
    Workspace,
    ProjectContainer,
    VfsMount,
    StoryContainer,
    AgentKnowledge,
    Lifecycle,
    Canvas,
    ExternalService,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, TS, PartialEq, Eq, Default)]
pub struct ResolvedMountEditCapabilities {
    pub create: bool,
    pub delete: bool,
    pub rename: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS, PartialEq, Eq)]
pub struct ResolvedMountSummary {
    pub id: String,
    pub display_name: String,
    pub provider: String,
    pub backend_id: String,
    pub capabilities: Vec<String>,
    pub default_write: bool,
    pub purpose: ResolvedMountPurpose,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(optional)]
    pub backend_online: Option<bool>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(optional)]
    #[ts(type = "number")]
    pub file_count: Option<usize>,
    pub edit_capabilities: ResolvedMountEditCapabilities,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS, PartialEq, Eq)]
pub struct ResolvedVfsSurface {
    pub surface_ref: String,
    pub source: ResolvedVfsSurfaceSource,
    pub mounts: Vec<ResolvedMountSummary>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(optional)]
    pub default_mount_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
pub struct ResolveSurfaceRequest {
    pub source: ResolvedVfsSurfaceSource,
}

#[derive(Debug, Clone, Deserialize, TS)]
pub struct SurfaceEntriesQuery {
    #[serde(default)]
    #[ts(optional)]
    pub path: Option<String>,
    #[serde(default)]
    #[ts(optional)]
    pub pattern: Option<String>,
    #[serde(default)]
    #[ts(optional)]
    pub recursive: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
pub struct SurfaceMountEntry {
    pub path: String,
    pub entry_type: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(optional)]
    #[ts(type = "number")]
    pub size: Option<u64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(optional)]
    pub content_kind: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(optional)]
    pub mime_type: Option<String>,
    pub is_dir: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
pub struct SurfaceEntriesResponse {
    pub surface_ref: String,
    pub mount_id: String,
    pub entries: Vec<SurfaceMountEntry>,
}

#[derive(Debug, Clone, Deserialize, TS)]
pub struct SurfaceReadFileRequest {
    pub surface_ref: String,
    pub mount_id: String,
    pub path: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
pub struct SurfaceReadFileResponse {
    pub surface_ref: String,
    pub mount_id: String,
    pub path: String,
    pub content: String,
    #[ts(type = "number")]
    pub size: u64,
    pub content_kind: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(optional)]
    pub mime_type: Option<String>,
}

#[derive(Debug, Clone, Deserialize, TS)]
pub struct SurfaceWriteFileRequest {
    pub surface_ref: String,
    pub mount_id: String,
    pub path: String,
    pub content: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
pub struct SurfaceWriteFileResponse {
    pub surface_ref: String,
    pub mount_id: String,
    pub path: String,
    #[ts(type = "number")]
    pub size: u64,
    pub persisted: bool,
    pub content_kind: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(optional)]
    pub mime_type: Option<String>,
}

#[derive(Debug, Clone, Deserialize, TS)]
pub struct SurfaceCreateFileRequest {
    pub surface_ref: String,
    pub mount_id: String,
    pub path: String,
    pub content: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
pub struct SurfaceCreateFileResponse {
    pub surface_ref: String,
    pub mount_id: String,
    pub path: String,
    #[ts(type = "number")]
    pub size: u64,
    pub content_kind: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(optional)]
    pub mime_type: Option<String>,
}

#[derive(Debug, Clone, Deserialize, TS)]
pub struct SurfaceDeleteFileRequest {
    pub surface_ref: String,
    pub mount_id: String,
    pub path: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
pub struct SurfaceDeleteFileResponse {
    pub surface_ref: String,
    pub mount_id: String,
    pub path: String,
    pub deleted: bool,
}

#[derive(Debug, Clone, Deserialize, TS)]
pub struct SurfaceRenameFileRequest {
    pub surface_ref: String,
    pub mount_id: String,
    pub from_path: String,
    pub to_path: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
pub struct SurfaceRenameFileResponse {
    pub surface_ref: String,
    pub mount_id: String,
    pub from_path: String,
    pub to_path: String,
}

#[derive(Debug, Clone, Deserialize, TS)]
pub struct SurfaceStatFileRequest {
    pub surface_ref: String,
    pub mount_id: String,
    pub path: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
pub struct SurfaceStatFileResponse {
    pub surface_ref: String,
    pub mount_id: String,
    pub path: String,
    pub entry_type: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(optional)]
    #[ts(type = "number")]
    pub size: Option<u64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(optional)]
    pub content_kind: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(optional)]
    pub mime_type: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(optional)]
    #[ts(type = "number")]
    pub modified_at: Option<i64>,
    pub is_dir: bool,
}

#[derive(Debug, Clone, Deserialize, TS)]
pub struct SurfaceApplyPatchRequest {
    pub surface_ref: String,
    pub mount_id: String,
    pub patch: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
pub struct SurfaceApplyPatchResponse {
    pub surface_ref: String,
    pub mount_id: String,
    pub added: Vec<String>,
    pub modified: Vec<String>,
    pub deleted: Vec<String>,
}

#[derive(Debug, Clone, Deserialize, TS)]
pub struct SurfaceReadBinaryFileRequest {
    pub surface_ref: String,
    pub mount_id: String,
    pub path: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
pub struct SurfaceUploadBinaryFileResponse {
    pub surface_ref: String,
    pub mount_id: String,
    pub path: String,
    #[ts(type = "number")]
    pub size: u64,
    pub content_kind: String,
    pub mime_type: String,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, TS, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum VfsCapabilityDto {
    Read,
    Write,
    List,
    Search,
    Exec,
    /// 订阅内容变更事件。
    Watch,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS, PartialEq, Eq)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum ProjectVfsMountContentDto {
    Inline,
    ExternalService {
        service_id: String,
        root_ref: String,
    },
}

#[derive(Debug, Clone, Deserialize, TS)]
pub struct CreateProjectVfsMountRequest {
    pub mount_id: String,
    pub display_name: String,
    #[serde(default)]
    #[ts(optional)]
    pub description: Option<String>,
    #[serde(default)]
    pub capabilities: Vec<VfsCapabilityDto>,
    pub content: ProjectVfsMountContentDto,
}

pub type UpdateProjectVfsMountRequest = CreateProjectVfsMountRequest;

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
pub struct ProjectVfsMountResponse {
    pub project_id: String,
    pub mount_id: String,
    pub display_name: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(optional)]
    pub description: Option<String>,
    pub capabilities: Vec<VfsCapabilityDto>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(optional)]
    pub installed_source: Option<InstalledAssetSourceResponse>,
    pub content: ProjectVfsMountContentDto,
    pub surface_ref: String,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS, PartialEq, Eq)]
pub struct InstalledAssetSourceResponse {
    pub library_asset_id: String,
    pub source_ref: String,
    pub source_version: String,
    pub source_digest: String,
    pub installed_at: String,
}
